main.py

```diff
@@ -60,38 +60,56 @@ def run_nn_policy(env, model):
       second number is the total number of actions taken before the
       episode finished.
     """
-    state = env.reset()
+    old_state = env.reset()
     env.render()
 
+    episode = []
     total_reward = 0
     num_steps = 0
     while True:
-        state, reward, is_terminal, debug_info = env.step(
-            interface.build_nn_output(model.predict_on_batch(interface.build_nn_input(state, 2))))
+        state_rep = interface.build_nn_input(old_state, 2)
+        pred = model.predict_on_batch(state_rep)
+        action = interface.build_nn_output(pred, std_x=stddev, std_y=stddev)
+        new_state, reward, is_terminal, debug_info = env.step(action)
+        episode.append((state_rep, pred, action, reward))
         env.render()
+        old_state = new_state
 
         total_reward += reward
         num_steps += 1
 
         if is_terminal:
             break
-    return total_reward, num_steps
-
-
+    return total_reward, num_steps, episode
+
+def reinforce(env, model, episode, total_reward, stddev=1.0):
+    gt = total_reward
+    var = stddev ** 2
+    for state_rep, pred, action, reward in episode:
+        action_t = action.transpose()
+        target = (action_t - pred)/var * gt + pred
```

Comment:
Since we are doing MSE with 2 dimensions the loss function here is 1/2 (mean_x - target_x)^2 + 1/2 (mean_y - target_y)^2. The factor of 2 gets cancelled out. Actually, I was talking with Ananya and the reason this works at all is a little more complicated then we originally thought. I can explain at next meeting. Also, the algorithm I'm using is the one on page 269 of Sutton and Barto, which uses G_t, the cumulative return from step t. If this is not the right algorithm let me know.